yjit/src/codegen.rs

```diff
@@ -662,31 +662,33 @@ pub fn gen_entry_prologue(cb: &mut CodeBlock, ocb: &mut OutlinedCb, iseq: IseqPt
     asm.cpush(EC);
     asm.cpush(SP);
 
-    // We are passed EC and CFP as arguments
+    // We are passed EC, CFP, and handle_exception flag as arguments
     asm.mov(EC, C_ARG_OPNDS[0]);
     asm.mov(CFP, C_ARG_OPNDS[1]);
+    let handle_exception = C_ARG_OPNDS[2];
 
     // Load the current SP from the CFP into REG_SP
     asm.mov(SP, Opnd::mem(64, CFP, RUBY_OFFSET_CFP_SP));
 
     // Setup cfp->jit_return
-    asm.mov(
-        Opnd::mem(64, CFP, RUBY_OFFSET_CFP_JIT_RETURN),
-        Opnd::const_ptr(CodegenGlobals::get_leave_exit_code().raw_ptr()),
+    asm.ccall(
+        rb_yjit_set_jit_return as _,
+        vec![
+            CFP,
+            CodegenGlobals::get_leave_compile_code().into_usize().into(),
+            CodegenGlobals::get_leave_exit_code().into_usize().into(),
+            handle_exception,
+        ],
```

Comment:
I don't love that you're adding an extra parameter to every JIT entry point and also calling a C function in there... We want entry points to be as streamlined as possible.

Could you compile a different kind of entry point if `handle_exception` is true?

Reply:
> Could you compile a different kind of entry point if handle_exception is true?

Yeah, I think we could. We could let each ISEQ have two function pointers instead of one, e.g. `jit_func` and `jit_exception`. `jit_exec(ec, false)` calls `jit_func`, and `jit_exec(ec, true)` calls `jit_exception`. Then `jit_func` can be as streamlined as it currently is. `jit_exception` still needs to be complicated, but at least this would leave existing paths intact.

I hesitated to do that because that would complicate the implementation a bit, but I'll work on it to streamline `jit_func`, limiting the complex setup to `jit_exception` paths.

Reply:
But will each ISEQ be guaranteed to have only one entry point for handling exceptions? Can't exceptions be caught at multiple different locations in the ISEQ?

Also it's not that great if there is still a conditional check in `jit_exec`. Hmmm. Would it be possible to separate out the logic for handling exections into a function separate from `jit_exec`? 😅  That way there would be no boolean argument and no if statement for this?

Reply:
Hopefully, I think my proposed solution actually addresses all of your concerns 🙂  I'll try to push that as soon as possible.

Reply:
This is what I was thinking of https://github.com/ruby/ruby/pull/8171/commits/5d79afbe66435e9a4735cc1bcc31814db39c503e. With this patch, existing `jit_exec` behaviors do not change.